go/vt/vtgate/engine/route.go

```diff
@@ -431,10 +436,57 @@ func (route *Route) execUpdateEqual(vcursor VCursor, bindVars map[string]*queryp
 	if len(ksid) == 0 {
 		return &sqltypes.Result{}, nil
 	}
+	if len(route.ChangedVindexValues) != 0 {
+		result, err := route.execUpdateEqualChangedVindex(vcursor, route.Subquery, bindVars, ks, shard, ksid)
+		if err != nil {
+			return nil, err
+		}
+		return result, nil
+	}
 	rewritten := sqlannotation.AddKeyspaceIDs(route.Query, [][]byte{ksid}, "")
 	return route.execShard(vcursor, rewritten, bindVars, ks, shard, true /* isDML */)
 }
 
+// execUpdateEqualChangedVindex performs an update when a vindex is being modified
+// by the statement.
+// Note: The order seen in the DML's seen below won't neccesarly matches
+// the one that will be executed by vttablet.
+// Given that dmls will be mapped to existent transactions on each
+// shard, the order could change.
+// However, to avoid issues with duplicate key constraints violations in the
+// case that a vindex update lands in the same shard, it's more convenient
+// to do the delete first. The following is an example of this sceneario:
+// - An update statement with an unique vindex column that sets a value to
+//   to what already exists in the db:
+//   update user set name='juan' where user_id=1
+//   Followed by:
+//   update user set name='juan' where user_id=1
+// If we don't perform the delete first, this query will fail because the lookup
+// table already has an entry from name=juan to user_id=1.
+//
+// Note 2: While changes are being committed, the changing row could be
+// unreachable by either the new or old column values.
+func (route *Route) execUpdateEqualChangedVindex(vcursor VCursor, query string, bindVars map[string]*querypb.BindVariable, keyspace, shard string, keyspaceID []byte) (*sqltypes.Result, error) {
+	var subQueryResult *sqltypes.Result
+	var err error
+	rewritten := sqlannotation.AddKeyspaceIDs(route.Query, [][]byte{keyspaceID}, "")
```

Comment:
Strictly speaking, this is only needed for statements that will make it to the binlog, and this one won't.
But then, we might have added it for selects in other places already.